### PR TITLE
[HEISENBUG] Safeguard color and alpha against mysterious IndexError

### LINKED ISSUE #908: Specviz: load data in multiple instances causes traceback in notebook
**Reporter: Camilla Pacifici**

If I open an instances of specviz in the notebook (viz), load a spectrum, then open a new instance (viz2), open a spectrum in viz2, then open another spectrum in viz, I get this traceback:
```
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
/tmp/ipykernel_183/2986648741.py in <module>
      3 template_cut = Spectrum1D(spectral_axis=template.spectral_axis[use_tmp], flux=template.flux[use_tmp])
      4 
----> 5 viz.load_spectrum(template_cut)

/opt/conda/envs/jwebbinar/lib/python3.8/site-packages/jdaviz/configs/specviz/helper.py in load_spectrum(self, data, data_label, format, show_in_viewer)
     25 
     26     def load_spectrum(self, data, data_label=None, format=None, show_in_viewer=True):
---> 27         super().load_data(data,
     28                           'specviz-spectrum1d-parser',
     29                           data_label=data_label,

/opt/conda/envs/jwebbinar/lib/python3.8/site-packages/jdaviz/core/helpers.py in load_data(self, data, parser_reference, **kwargs)
     44 
     45     def load_data(self, data, parser_reference=None, **kwargs):
---> 46         self.app.load_data(data, parser_reference=parser_reference, **kwargs)
     47 
     48     @property

/opt/conda/envs/jwebbinar/lib/python3.8/site-packages/jdaviz/app.py in load_data(self, file_obj, parser_reference, **kwargs)
    335                 # If the parser returns something other than known, assume it's
    336                 #  a message we want to make the user aware of.
--> 337                 msg = parser(self, file_obj, **kwargs)
    338 
    339                 if msg is not None:

/opt/conda/envs/jwebbinar/lib/python3.8/site-packages/jdaviz/configs/specviz/plugins/parsers.py in specviz_spectrum1d_parser(app, data, data_label, format, show_in_viewer)
     81             # Only auto-show the first spectrum in a list
     82             if i == 0 and show_in_viewer:
---> 83                 app.add_data_to_viewer("spectrum-viewer", data_label[i])

/opt/conda/envs/jwebbinar/lib/python3.8/site-packages/jdaviz/app.py in add_data_to_viewer(self, viewer_reference, data_path, clear_other_data, ext)
    708         if data_id is not None:
    709             data_ids.append(data_id)
--> 710             self._update_selected_data_items(viewer_item['id'], data_ids)
    711         else:
    712             raise ValueError(

/opt/conda/envs/jwebbinar/lib/python3.8/site-packages/jdaviz/app.py in _update_selected_data_items(self, viewer_id, selected_items)
   1018             [data] = [x for x in self.data_collection if x.label == label]
   1019 
-> 1020             viewer.add_data(data)
   1021 
   1022             add_data_message = AddDataMessage(data, viewer,

/opt/conda/envs/jwebbinar/lib/python3.8/site-packages/jdaviz/configs/specviz/plugins/viewers.py in add_data(self, data, color, alpha, **layer_state)
    336             self._color = color
    337 
--> 338         self._alpha = self.figure.marks[self.data_trace_pointer].opacities[0]
    339         if alpha:
    340             self._alpha = alpha

IndexError: list index out of range
```

[🐱](https://jira.stsci.edu/browse/JDAT-1790)

 --- 

 DISCLAIMER: This issue was autocreated by the Jdaviz Issue Creation Bot on behalf of the reporter. If any information is incorrect, please contact [Duy Nguyen](https://github.com/duytnguyendtn)

### PR DESCRIPTION
**Pull request by pllim**
### Description

This pull request is to theoretically safeguard against mysterious `IndexError` that I cannot reproduce in Specviz.

Since I cannot reproduce it myself, I am not sure what to write in the change log. I guess we don't need one? I will add one if you disagree.

Fixes #908

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [ ] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [x] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Did the CI pass? If not, are the failures related?
- [x] Is a change log needed? If yes, is it added to `CHANGES.rst`?
- [x] Is a milestone set? Milestone is only currently required for PRs related to Imviz MVP.
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)?

### CODE REVIEW

File: jdaviz/configs/specviz/plugins/viewers.py [hunk 1]
```diff
@@ -344,13 +344,22 @@ def add_data(self, data, color=None, alpha=None, **layer_state):
...
         # Color and opacity are taken from the already plotted trace,
         # in case they are not set explicitly by the caller.
-        self._color = self.figure.marks[self.data_trace_pointer].colors[0]
         if color:
             self._color = color
+        else:
+            try:
+                self._color = self.figure.marks[self.data_trace_pointer].colors[0]
+            except IndexError:  # Just grab from bqplot default
+                from bqplot.colorschemes import CATEGORY10
```
Comment (pllim) [inline jdaviz/configs/specviz/plugins/viewers.py:353]: Ref: https://github.com/bqplot/bqplot/blob/88ef9769594b9aa20162f61ed462d06bd06e72fe/bqplot/marks.py#L301

File: jdaviz/configs/specviz/plugins/viewers.py [hunk 2]
```diff
@@ -344,13 +344,22 @@ def add_data(self, data, color=None, alpha=None, **layer_state):
...
+                self._color = CATEGORY10[0]  # Should be #1f77b4 (bluish)
 
-        self._alpha = self.figure.marks[self.data_trace_pointer].opacities[0]
         if alpha:
             self._alpha = alpha
+        else:
+            try:
+                self._alpha = self.figure.marks[self.data_trace_pointer].opacities[0]
```
Comment (pllim) [inline jdaviz/configs/specviz/plugins/viewers.py:360]: Ref: https://github.com/bqplot/bqplot/blob/88ef9769594b9aa20162f61ed462d06bd06e72fe/bqplot/marks.py#L311

**Comment by kecnry** [comment]
Although I'm all for fallbacks instead of tracebacks for something like this if we really can't track it down and fix the underlying cause, will it ever result in inconsistent or unexpected behavior (in which case maybe a snackbar warning would be helpful)?  Is there any concern that the underlying index error will cause other (larger or harder to detect) issues if left "ignored"?

**Comment by pllim** [comment]
@kecnry , they are all good questions! We can also just keep this PR open and revisit when someone sees the bug again consistently.

As for snackbar warning, I am not convinced. In principle, I don't see much use in warning users of things they cannot fix themselves. They will just end up ignoring the warnings and getting annoyed by all the notifications.

**Comment by pllim** [comment]
I am just going to close this without merge for now. If this comes up again, we can revisit.

**Pull request closed without merging**